File: algorithms/java/乘积最大子数组/乘积最大子数组.java
// Source : https://leetcode.cn/problems/maximum-product-subarray/
// Author : Yanan Wang
// Date   : 2023-06-23

class Solution {
    public int maxSubarraySumCircular(int[] nums) {
        int n = nums.length;    
        int[][] dp = new int[n + 1][2];
        int sum = 0;
        int max = nums[0];
        int min = nums[0];
        for (int i = 1; i <= n; i++) {
            dp[i][0] = Math.max(nums[i - 1], dp[i - 1][0] + nums[i - 1]);
            dp[i][1] = Math.min(nums[i - 1], dp[i - 1][1] + nums[i - 1]);
            sum += nums[i - 1];
            max = Math.max(max, dp[i][0]);
            min = Math.min(min, dp[i][1]);
        }
        if (max < 0) {
            return max;
        }
        return Math.max(max, sum - min);
    }
}

File: algorithms/python/乘积最大子数组/乘积最大子数组.py
# Source : https://leetcode.cn/problems/maximum-product-subarray/description/
# Author : Yanan Wang
# Date   : 2024-01-16

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n = len(nums)
        max_dp = [1] * (n + 1) 
        min_dp = [1] * (n + 1) 
        ans = nums[0]

        for i in range(1, n + 1):
            max_dp[i] = max(nums[i - 1], max_dp[i - 1] * nums[i - 1], min_dp[i - 1] * nums[i - 1])
            min_dp[i] = min(nums[i - 1], max_dp[i - 1] * nums[i - 1], min_dp[i - 1] * nums[i - 1])
            ans = max(ans, max_dp[i])

        return ans